include/spark_dsg/traversability_boundary.h

```diff
@@ -0,0 +1,411 @@
+ *     this list of conditions and the following disclaimer.
+ *
+ *  2. Redistributions in binary form must reproduce the above copyright notice,
+ *     this list of conditions and the following disclaimer in the documentation
+ *     and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Research was sponsored by the United States Air Force Research Laboratory and
+ * the United States Air Force Artificial Intelligence Accelerator and was
+ * accomplished under Cooperative Agreement Number FA8750-19-2-1000. The views
+ * and conclusions contained in this document are those of the authors and should
+ * not be interpreted as representing the official policies, either expressed or
+ * implied, of the United States Air Force or the U.S. Government. The U.S.
+ * Government is authorized to reproduce and distribute reprints for Government
+ * purposes notwithstanding any copyright notation herein.
+ * -------------------------------------------------------------------------- */
+#pragma once
+
+#include <spark_dsg/node_attributes.h>
+
+#include <Eigen/Geometry>
+
+namespace spark_dsg {
+
+/**
+ * @brief Check whether a state is considered traversable, i.e., it is TRAVERSABLE or
+ * TRAVERSED. If optimistic is true, UNKNOWN is also considered traversable.
+ */
+bool isTraversable(TraversabilityState state, bool optimistic = false);
+
+/** Check whether a set of states is considered traversable, i.e., all states are
+ * TRAVERSABLE or TRAVERSED. If optimistic is true, UNKNOWN is also considered
+ * traversable.
+ */
+bool isTraversable(const TraversabilityStates& states, bool optimistic = false);
+
+/**
+ * @brief Check whether a state is considered intraversable, i.e., it is INTRAVERSABLE.
+ * If optimistic is false, UNKNOWN is also considered intraversable.
+ */
+bool isIntraversable(TraversabilityState state, bool optimistic = false);
+
+/** Check whether a set of states is considered intraversable, i.e., at least one state
+ * is INTRAVERSABLE. If optimistic is false, UNKNOWN is also considered intraversable.
+ */
+bool isIntraversable(const TraversabilityStates& states, bool optimistic = false);
```

Comment:
I sorta see the point for having both `isTraversable` and `isIntraversable`, but think `isIntraversable` does actually end up working out to be `!isTraversable`? Also the difference between State and States is subtle, I would do `areAllTraversable` instead of the overload